driver/src/orderbook/streamed/state.rs

```diff
@@ -787,4 +787,106 @@ mod tests {
         assert_eq!(balance.get(&(address(2), 0)), Some(&12.into()));
         assert_eq!(balance.get(&(address(2), 1)), Some(&9.into()));
     }
+
+    #[test]
+    fn orderbook_solution_touching_same_order_twice() {
+        let mut state = state_with_fee();
+        let event = TokenListing {
+            token: address(1),
+            id: 1,
+        };
+        state = state.apply_event(&Event::TokenListing(event), 0).unwrap();
+
+        let event = Deposit {
+            user: address(2),
+            token: address(0),
+            amount: 20.into(),
+            batch_id: 0,
+        };
+        state = state.apply_event(&Event::Deposit(event), 0).unwrap();
+        let event = Deposit {
+            user: address(3),
+            token: address(1),
+            amount: 20.into(),
+            batch_id: 0,
+        };
+        state = state.apply_event(&Event::Deposit(event), 0).unwrap();
+
+        let event = OrderPlacement {
+            owner: address(2),
+            index: 0,
+            buy_token: 1,
+            sell_token: 0,
+            valid_from: 0,
+            valid_until: 10,
+            price_numerator: 10,
+            price_denominator: 10,
+        };
+        state = state.apply_event(&Event::OrderPlacement(event), 0).unwrap();
+        let mut event = OrderPlacement {
+            owner: address(3),
+            index: 0,
+            buy_token: 0,
+            sell_token: 1,
+            valid_from: 0,
+            valid_until: 10,
+            price_numerator: 5,
+            price_denominator: 5,
+        };
+        state = state
+            .apply_event(&Event::OrderPlacement(event.clone()), 0)
+            .unwrap();
+        event.index = 1;
```

Comment:
Why do we need two order placements here? It feels like one counter order should be enough.

Reply:
How can you match an order without a counter order?

Reply:
Yes, a single order would have been enough. I went for two orders because:
1. they serve as a contrast to what is tested in this PR (a single order split into two should be like two independent orders)
2. I see it slightly more likely to fail when the same order is matched against two different orders than when matched twice against the same order.